### PR TITLE
using controlled phase grad

### PR DESCRIPTION
**Pull request by skushnir123**
Small PR to now use a controlled version of AddIntoPhaseGrad

### CODE REVIEW

File: qualtran/bloqs/rotations/phase_gradient.py [hunk 1]
```diff
@@ -183,21 +197,28 @@ def scaled_val(self, x: int) -> int:
         x_fxp = _fxp(x / 2**x_width, x_width).like(_fxp(0, self.phase_bitsize)).astype(float)
         return int(x_fxp.astype(float) * 2**self.phase_bitsize)
 
-    def apply(self, x: int, phase_grad: int) -> Union[int, Iterable[int]]:
-        out = self.on_classical_vals(x=x, phase_grad=phase_grad)
-        return out['x'], out['phase_grad']
+    def apply(self, ctrl: int, x: int, phase_grad: int) -> Union[int, Iterable[int]]:
```
Comment (tanujkhattar) [inline qualtran/bloqs/rotations/phase_gradient.py:200]: You'll have to change this to `*args` because `apply` will be called with only 2 parameters with `self.control` is False.

File: qualtran/bloqs/rotations/phase_gradient.py [hunk 2]
```diff
@@ -150,27 +152,39 @@ class AddIntoPhaseGrad(GateWithRegisters, cirq.ArithmeticGate):
         [Compilation of Fault-Tolerant Quantum Heuristics for Combinatorial Optimization](https://arxiv.org/abs/2007.07391)
         Appendix A: Addition for controlled rotations
     """
+
     x_bitsize: int
     phase_bitsize: int
     right_shift: int = 0
     sign: int = +1
+    controlled: bool = False
```
Comment (tanujkhattar) [inline qualtran/bloqs/rotations/phase_gradient.py:160]: We've used `controlled: Optional[bool] = None` in the past to denote 3 cases
a) None -  no controls
b) 1 - Controlled with value 1
c) 0 - Controlled with value 0

File: qualtran/bloqs/rotations/phase_gradient.py [hunk 3]
```diff
@@ -183,21 +197,28 @@ def scaled_val(self, x: int) -> int:
...
-        out = self.on_classical_vals(x=x, phase_grad=phase_grad)
-        return out['x'], out['phase_grad']
+    def apply(self, ctrl: int, x: int, phase_grad: int) -> Union[int, Iterable[int]]:
+        out = self.on_classical_vals(ctrl=ctrl, x=x, phase_grad=phase_grad)
+        return out['ctrl'], out['x'], out['phase_grad']
 
-    def on_classical_vals(self, x: int, phase_grad: int) -> Dict[str, 'ClassicalValT']:
-        phase_grad_out = (phase_grad + self.sign * self.scaled_val(x)) % (2**self.phase_bitsize)
-        return {'x': x, 'phase_grad': phase_grad_out}
+    def on_classical_vals(self, ctrl: int, x: int, phase_grad: int) -> Dict[str, 'ClassicalValT']:
```
Comment (tanujkhattar) [inline qualtran/bloqs/rotations/phase_gradient.py:204]: ditto

File: qualtran/bloqs/rotations/phase_gradient.py [hunk 4]
```diff
@@ -183,21 +197,28 @@ def scaled_val(self, x: int) -> int:
...
+        )
+        return {'ctrl': ctrl, 'x': x, 'phase_grad': phase_grad_out}
 
     def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
-        num_toffoli = self.phase_bitsize - 2
-        return {(Toffoli(), num_toffoli)}
+        if self.controlled == False:
+            num_toffoli = self.phase_bitsize - 2
+            return {(Toffoli(), num_toffoli)}
+        return {(TGate(), 8 * self.phase_bitsize)}
```
Comment (tanujkhattar) [inline qualtran/bloqs/rotations/phase_gradient.py:214]: If `num_toffoli = self.phase_bitsize-2`, the call graph should decompose into `num_toffoli` $\text{And}$ and $\text{And}^\dagger$'s for the non-controlled case and `2 * num_toffoli` of each for the controlled case. 


The fact that we need twice the number of $\text{And}$ / $\text{And}^\dagger$ comes from https://arxiv.org/pdf/1709.06648.pdf (see primitives for non-controlled additions and controlled additions below)


image


image


And the fact that we need `self.phase_bitsize - 2` instead of `self.phase_bitsize - 1` comes from the optimization described in Fig 13 of https://arxiv.org/pdf/2007.07391.pdf where they get rid of the last carry bit utilizing the fact that target register is a phase gradient register

File: qualtran/bloqs/rotations/phase_gradient.py [hunk 5]
```diff
@@ -50,6 +51,7 @@ class PhaseGradientUnitary(GateWithRegisters):
 
     The implementation simply decomposes into $n$ (controlled-) rotations, one on each qubit.
     """
+
```
Comment (tanujkhattar) [inline qualtran/bloqs/rotations/phase_gradient.py:54]: Revert unrelated change

File: qualtran/bloqs/rotations/phase_gradient.py [hunk 6]
```diff
@@ -150,27 +152,39 @@ class AddIntoPhaseGrad(GateWithRegisters, cirq.ArithmeticGate):
         [Compilation of Fault-Tolerant Quantum Heuristics for Combinatorial Optimization](https://arxiv.org/abs/2007.07391)
         Appendix A: Addition for controlled rotations
     """
+
     x_bitsize: int
     phase_bitsize: int
     right_shift: int = 0
     sign: int = +1
+    controlled: Optional[bool] = None
```
Comment (tanujkhattar) [inline qualtran/bloqs/rotations/phase_gradient.py:160]: Please update the docstring to include the `controlled` arg in the `Args` and `Registers` section. Also change to `controlled: Optional[int] = None` s.t. `controlled=0` means control should be active when ctrl register stores value `0`. I think right now, your implementation below works only for the case when control register stores value `1` (i.e. controlled=True)

File: qualtran/bloqs/rotations/phase_gradient.py [hunk 7]
```diff
@@ -183,16 +197,33 @@ def scaled_val(self, x: int) -> int:
...
         return int(x_fxp.astype(float) * 2**self.phase_bitsize)
 
-    def apply(self, x: int, phase_grad: int) -> Union[int, Iterable[int]]:
+    def apply(self, *args) -> Union[int, Iterable[int]]:
+        if self.controlled:
+            ctrl, x, phase_grad = args
+            out = self.on_classical_vals(ctrl=ctrl, x=x, phase_grad=phase_grad)
+            return out['ctrl'], out['x'], out['phase_grad']
+
+        x, phase_grad = args
```
Comment (tanujkhattar) [inline qualtran/bloqs/rotations/phase_gradient.py:206]: For all the changes made to `AddIntoPhaseGrad`, you should add tests to `phase_gradient_test.py` to verify the correctness of these changes.

File: qualtran/bloqs/rotations/phase_gradient.py [hunk 8]
```diff
@@ -183,16 +197,33 @@ def scaled_val(self, x: int) -> int:
...
+            )
+            return {'ctrl': ctrl, 'x': x, 'phase_grad': phase_grad_out}
+
         phase_grad_out = (phase_grad + self.sign * self.scaled_val(x)) % (2**self.phase_bitsize)
         return {'x': x, 'phase_grad': phase_grad_out}
 
     def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
         num_toffoli = self.phase_bitsize - 2
+        if self.controlled:
+            return {(TGate(), 2 * num_toffoli)}
```
Comment (tanujkhattar) [inline qualtran/bloqs/rotations/phase_gradient.py:225]:
```suggestion
            return {(Toffoli(), 2 * num_toffoli)}
```

File: qualtran/bloqs/qft/approximate_qft_test.py [hunk 9]
```diff
@@ -70,7 +69,6 @@ def test_approximate_qft_exact(n: int, without_reverse: bool):
     qft_bloq = TestApproximateQFT(n, n, not without_reverse)
     qft_cirq = cirq.QuantumFourierTransformGate(n, without_reverse=without_reverse)
     np.testing.assert_allclose(cirq.unitary(qft_bloq), cirq.unitary(qft_cirq))
-    np.testing.assert_allclose(cirq.unitary(qft_bloq**-1), cirq.unitary(qft_cirq**-1))
```
Comment (tanujkhattar) [inline qualtran/bloqs/qft/approximate_qft_test.py:73]: Why was this removed? Does this assertion not pass anymore?

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
Please add more tests

File: qualtran/bloqs/qft/approximate_qft_test.py (hunk 9 above)
Comment (skushnir123) [reply qualtran/bloqs/qft/approximate_qft_test.py:73]: added back after adding inverse implementation for AddIntoPhaseGrad

File: qualtran/bloqs/rotations/phase_gradient.py (hunk 5 above)
Comment (skushnir123) [reply qualtran/bloqs/rotations/phase_gradient.py:54]: I think this was the automatic formatter

Comment (skushnir123) [reply qualtran/bloqs/rotations/phase_gradient.py:54]: Should I still revert?

Comment (tanujkhattar) [reply qualtran/bloqs/rotations/phase_gradient.py:54]: nope, that's fine.
